File: packages/septr-go/handshake_test.go
package septr

import (
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
)

func TestSendHandshake(t *testing.T) {
	var gotBody map[string]interface{}
	var gotAuth string

	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/handshake" {
			t.Errorf("expected /handshake, got %s", r.URL.Path)
		}
		gotAuth = r.Header.Get("Authorization")
		if err := json.NewDecoder(r.Body).Decode(&gotBody); err != nil {
			t.Fatalf("decode body: %v", err)
		}
		w.Header().Set("Content-Type", "application/json")
		w.Write([]byte(`{"status":"connected","project":{"id":"p1","name":"Demo"}}`))
	}))
	defer server.Close()

	url := strings.TrimSuffix(server.URL, "/") + "/events"
	config := &Config{APIKey: "septr_live_11111111-2222-3333-4444-555555555555_abcd0123456789abcdef0123456789ab", TelemetryURL: url, framework: "gin"}

	if ok := sendHandshake(config, config.APIKey); !ok {
		t.Fatal("expected handshake to succeed")
	}
	if gotAuth != "Bearer "+config.APIKey {
		t.Errorf("expected bearer auth, got %q", gotAuth)
	}
	if gotBody["runtime"] != "gin" {
		t.Errorf("expected runtime gin, got %v", gotBody["runtime"])
	}
	if gotBody["package"] != "septr" {
		t.Errorf("expected package septr, got %v", gotBody["package"])
	}
	if gotBody["version"] != sdkVersion {
		t.Errorf("expected version %s, got %v", sdkVersion, gotBody["version"])
	}
}

func TestSendHandshakeRejected(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusUnauthorized)
		w.Write([]byte(`{"detail":"Invalid API key"}`))
	}))
	defer server.Close()

	config := &Config{APIKey: "bad", TelemetryURL: server.URL + "/events"}
	if ok := sendHandshake(config, config.APIKey); ok {
		t.Fatal("expected handshake to fail on 401")
	}
}

func TestSendHandshakeNoKey(t *testing.T) {
	config := &Config{}
	if ok := sendHandshake(config, ""); ok {
		t.Fatal("expected handshake to fail without a key")
	}
}

func TestHandshakeURL(t *testing.T) {
	cases := []struct {
		telemetryURL string
		want         string
	}{
		{"https://api.septr.com/v1/events", "https://api.septr.com/v1/handshake"},
		{"http://localhost:8000/v1/events", "http://localhost:8000/v1/handshake"},
		{"http://localhost:8000/events", "http://localhost:8000/handshake"},
		{"", "https://api.septr.com/v1/handshake"},
	}
	for _, c := range cases {
		config := &Config{TelemetryURL: c.telemetryURL}
		if got := handshakeURL(config); got != c.want {
			t.Errorf("handshakeURL(%q) = %q, want %q", c.telemetryURL, got, c.want)
		}
	}
}

File: packages/septr-node/src/__tests__/handshake.test.ts
import { describe, it, expect, vi, afterEach } from "vitest"
import { sendHandshake } from "../core/telemetry"

afterEach(() => {
  vi.restoreAllMocks()
})

const KEY = "septr_live_11111111-2222-3333-4444-555555555555_abcd0123456789abcdef0123456789ab"

describe("sendHandshake", () => {
  it("posts runtime/package/version and returns true on connected", async () => {
    const fetchMock = vi.fn(async () => {
      return {
        ok: true,
        status: 200,
        json: async () => ({ status: "connected", project: { id: "p1", name: "Demo" } }),
      }
    })
    vi.stubGlobal("fetch", fetchMock)

    const ok = await sendHandshake({ apiKey: KEY, telemetryUrl: "https://api.septr.com/v1/events", framework: "express" })

    expect(ok).toBe(true)
    const [url, init] = fetchMock.mock.calls[0] as unknown as [string, RequestInit]
    const headers = init.headers as Record<string, string>
    expect(url).toBe("https://api.septr.com/v1/handshake")
    expect(headers.Authorization).toBe(`Bearer ${KEY}`)
    expect(JSON.parse(init.body as string)).toMatchObject({
      runtime: "express",
      package: "septr",
      version: expect.any(String),
      environment: "test",
    })
  })

  it("strips /events from custom telemetry URLs", async () => {
    const fetchMock = vi.fn(async () => {
      return {
        ok: true,
        status: 200,
        json: async () => ({ status: "connected", project: {} }),
      }
    })
    vi.stubGlobal("fetch", fetchMock)

    await sendHandshake({ apiKey: KEY, telemetryUrl: "http://localhost:8000/v1/events" })
    const [url] = fetchMock.mock.calls[0] as unknown as [string]
    expect(url).toBe("http://localhost:8000/v1/handshake")
  })

  it("returns false without a key", async () => {
    const ok = await sendHandshake({})
    expect(ok).toBe(false)
  })

  it("returns false on non-200", async () => {
    vi.stubGlobal("fetch", vi.fn(async () => ({ ok: false, status: 401 })))
    const ok = await sendHandshake({ apiKey: KEY })
    expect(ok).toBe(false)
  })

  it("returns false on network failure", async () => {
    vi.stubGlobal("fetch", vi.fn(async () => {
      throw new Error("boom")
    }))
    const ok = await sendHandshake({ apiKey: KEY })
    expect(ok).toBe(false)
  })
})
